src/sentry/utils/services.py

```diff
@@ -275,23 +296,25 @@ def execute(*args, **kwargs):
                     '{!r} is not a registered backend.'.format(
                         selected_backend_names[0]))
 
-            def call_backend_method(backend):
-                base = self.__backend_base
-                active_backends = type(self).state.backends
+            def call_backend_method(context, backend):
+                # Update the thread local state in the executor to the provided
+                # context object. This allows the context to be propagated
+                # across different threads.
+                assert type(self).__state.context is None
+                type(self).__state.context = context
 
                 # Ensure that we haven't somehow accidentally entered a context
                 # where the backend we're calling has already been marked as
                 # active (or worse, some other backend is already active.)
-                assert base not in active_backends
+                base = self.__backend_base
+                assert base not in context.backends
 
                 # Mark the backend as active.
-                active_backends[base] = backend
+                context.backends[base] = backend
                 try:
                     return getattr(backend, attribute_name)(*args, **kwargs)
                 finally:
-                    # Unmark the backend as active.
-                    assert active_backends[base] is backend
-                    del active_backends[base]
```

Comment:
This doesn't need to be explicitly deleted any longer since a copy was passed into this method.